src/libcore/slice/mod.rs

```diff
@@ -407,6 +407,65 @@ impl<T> [T] {
         self as *mut [T] as *mut T
     }
 
+    /// Returns the two raw pointers spanning the slice.
+    ///
+    /// The returned range is half-open, which means that the end pointer
+    /// points *one past* the last element of the slice. This way, an empty
+    /// slice is represented by two equal pointers, and the difference between
+    /// the two pointers represents the size of the size.
+    ///
+    /// See [`as_ptr`] for warnings on using these pointers. The end pointer
+    /// requires extra caution, as it does not point to a valid element in the
+    /// slice.
+    ///
+    /// This function is useful for interacting with foreign interfaces which
+    /// use two pointers to refer to a range of elements in memory, as is
+    /// common in C++.
+    ///
+    /// It can also be useful to check if a reference or pointer to an element
+    /// refers to an element of this slice:
+    ///
+    /// ```
+    /// let a = [1,2,3];
+    /// let x = &a[1];
+    /// let y = &5;
+    /// assert!(a.as_ptr_range().contains(x));
+    /// assert!(!a.as_ptr_range().contains(y));
+    /// ```
+    ///
+    /// [`as_ptr`]: #method.as_ptr
+    #[unstable(feature = "slice_ptr_range", issue = "0")]
+    #[inline]
+    pub fn as_ptr_range(&self) -> Range<*const T> {
+        let start = self.as_ptr();
+        let end = unsafe { start.add(self.len()) };
```

Comment:
Could you please provide a brief safety argument reasoning around the [requirements](https://doc.rust-lang.org/nightly/std/primitive.pointer.html#safety-4) `.add` states. (Notably, `vec.as_ptr().add(vec.len())` is mentioned.) (Same below.)

Reply:
Hm, now I'm starting to wonder if it actually *is* safe. A Vec is never larger than `isize::MAX` bytes, but that guarantee is not there for `[T]`, is it?

Reply:
A Vec is never larger than isize::MAX? Where did you get that information from?

Reply:
Also I don't see how that affects this. Even if Vec can only be a certain limited size, the pointers are not limited.

Reply:
Yes exactly, that's why I changed it to `wrapping_add` instead.

Reply:
Mmh, I'm wondering where this limitation with isize::MAX is coming from. The documentation doesn't seem to mention why that is, only that this is the case.

Reply:
cc @rust-lang/lang @rust-lang/wg-unsafe-code-guidelines

> Note that slices point to their entire range, so it is important that the length metadata is never too large. In particular, allocations and therefore slices cannot be bigger than `isize::MAX` bytes.

- https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html

> The total size of the slice must be no larger than `isize::MAX` bytes in memory. See the safety documentation of pointer::offset.

- https://doc.rust-lang.org/nightly/alloc/slice/fn.from_raw_parts.html#safety

(Assuming these statements are normative.)

Reply:
(In case it is not normative, an `assert!(...)`:
```rust
assert!(
    mem::size_of::<T>().saturating_mul(len) <= isize::MAX as usize,
    "attempt to create slice covering half the address space",
);
```
is another approach.)

Reply:
Seems it is in fact not guaranteed and under active discussion? cc https://github.com/rust-lang/unsafe-code-guidelines/issues/102#issuecomment-473340447.

Reply:
Hm, interesting. The first mentions allocations (does that mean heap allocations?), and the second one seems specific to `alloc`. Can't find a very clear statement saying that this holds for all slices. I'd expect something like that in the 'slice (primitive)' documentation itself.

However, compiling a static `[u8; 0x8000_0010]` on a 32-bit platform seems to suggest it applies there as well:

```
error[E0080]: the type `[u8; 2147483664]` is too big for the current architecture
 --> src/main.rs:1:1
  |
1 | static a: [u8; 0x8000_0010] = [0u8; 0x8000_0010];
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
Changed it back to `unsafe { add }`, with a comment describing why it should be safe.

Reply:
> The first mentions allocations (does that mean heap allocations?),

I think it refers to any sort of allocation, including stack allocations (as your array example indicates).

> and the second one seems specific to `alloc`.

That's a mistake on my part, it is defined in `core` as well.

-----------------------

So working on the assumption that this is unspecified / may become implementation-defined it seems to me that we have two choices:

1. Use `.wrapping_add(...)` -- potentially a foot-gun if you use `.contains` tho not going to happen for real programs because of LLVM limitations and because of the sheer size of the allocation.

2. Use `assert!(...)` as above. I'm not sure if LLVM would see the assertion and reason that it must always hold and so optimize it away.

We can probably switch between the two behaviors even after stabilizing because they are so pathological. However, it seems to me we should add a note to the documentation as well as the tracking issue and then the libs team can pick between one of them.

-----------------------

> Changed it back to `unsafe { add }`, with a comment describing why it should be safe.

Haha; Seems we had the opposite take-aways.

Reply:
Ah, the Index implementation of slices assume the same thing: https://doc.rust-lang.org/nightly/src/core/slice/mod.rs.html#2678-2724

Reply:
I don't want to wade into the issue of what's *normative*, but `ptr.offset` and related methods with restrictions to `isize::MAX` bytes are used pervasively from practically all safe slice APIs (at least, anything that ever accesses or takes the address of a slice element). Therefore:

- it is impossible in practice to have a *safe* slice (in the sense that you can do safe operations on it and rest assures they won't cause UB) that spans more than `isize::MAX` bytes
- we would have to audit and change a whole bunch of code throughout the standard library (or remove the `isize::MAX` restriction from `offset` and friends) if we ever wanted to support larger slices

So I think the current implementation (`unsafe { add }` with comment) is perfectly fine.´

I would advise against `wrapping_add` because it can cause performance degradation in some situations and there's no reason to risk that.

Reply:
Oh, I missed some of your comments. GitHub was not showing some.

Seeing that the Index implementation assumes the same thing, I figured it'd be best to just use the same.

Reply:
Yeah that seems like a good conclusion. Let's keep the PR as-is.

@m-ou-se Could you add a note to the second bullet re. this not being normative (at least not yet) as well as link to https://github.com/rust-lang/unsafe-code-guidelines/issues/102#issuecomment-473340447. With that I think we should be good to merge the PR.

(Also let's keep this thread visible since it was an interesting conversation.)

Reply:
Done.